File: Codex/Python/2025/Day6_Solution_2025.py
import time

from functools import reduce
from operator import mul


def main() -> None:
    with open("Day6_Input.txt", "r", encoding="utf-8") as file:
        lines = file.read().splitlines()

    width = max(len(line) for line in lines)
    padded = [line.ljust(width) for line in lines]

    separators = [all(line[col] == " " for line in padded) for col in range(width)]

    total = 0
    col = 0
    while col < width:
        while col < width and separators[col]:
            col += 1
        if col >= width:
            break

        start = col
        while col < width and not separators[col]:
            col += 1
        end = col

        values = [int(line[start:end].strip()) for line in padded[:-1]]
        operator = padded[-1][start:end].strip()

        if operator == "+":
            total += sum(values)
        else:
            total += reduce(mul, values, 1)

    print(total)


if __name__ == "__main__":
    start = time.perf_counter()
    main()
    end = time.perf_counter()
    print(f"Runtime: {end - start:.6f} seconds")

File: Codex/Rust/2025/Day6_Solution_2025.rs
use std::time::Instant;
use std::fs;

fn main() {
    let start = Instant::now();
    let input = fs::read_to_string("Day6_Input.txt").expect("failed to read Day6_Input.txt");
    let lines: Vec<&str> = input.lines().collect();
    if lines.is_empty() {
        println!("0");
        return;
    }

    let width = lines.iter().map(|line| line.len()).max().unwrap_or(0);
    let grid: Vec<Vec<u8>> = lines
        .iter()
        .map(|line| {
            let mut row = line.as_bytes().to_vec();
            row.resize(width, b' ');
            row
        })
        .collect();

    let mut total: u128 = 0;
    let mut col = 0usize;

    while col < width {
        let is_separator = grid.iter().all(|row| row[col] == b' ');
        if is_separator {
            col += 1;
            continue;
        }

        let start = col;
        while col < width && grid.iter().any(|row| row[col] != b' ') {
            col += 1;
        }
        let end = col;

        let mut numbers = Vec::new();
        for row in &grid[..grid.len() - 1] {
            let text = std::str::from_utf8(&row[start..end]).expect("invalid UTF-8").trim();
            if !text.is_empty() {
                numbers.push(text.parse::<u128>().expect("invalid number"));
            }
        }

        let op = std::str::from_utf8(&grid[grid.len() - 1][start..end])
            .expect("invalid UTF-8")
            .trim();

        let value = match op {
            "+" => numbers.into_iter().sum::<u128>(),
            "*" => numbers.into_iter().product::<u128>(),
            _ => panic!("invalid operator"),
        };

        total += value;
    }

    println!("{total}");

    println!("Runtime: {:.6} seconds", start.elapsed().as_secs_f64());
}
